File: python/xfa/08_picture_clause_formatting/08_picture_clause_formatting.py
# ============================================================================
#  08_picture_clause_formatting -- Python (ctypes) port of
#  examples\delphi\xfa\08_picture_clause_formatting\08_picture_clause_formatting.dpr
#
#  LumasPDF XFA dynamic engine "flavor tour" example 8 of 10 -- PICTURE-CLAUSE
#  FORMATTING. Demonstrates the <format><picture> formatter: real num{} /
#  date{} / text{} picture patterns applied both to plain bound data values
#  and to a value produced by a FormCalc <calculate> script (GrandTotalField),
#  proving the calculate-then-format pipeline order.
#
#    pdfNewPDF -> pdfCreateNewPDFA -> pdfCreateXFAStreamA('template',...) ->
#    pdfCreateXFAStreamA('datasets',...) -> pdfRenderXFAForm -> pdfCloseFile
#
#  Packets are read pre-split from their own .template.xml/.datasets.xml
#  files -- no XML library needed in this driver.
# ============================================================================
import os
import sys
# the wrapper lives at <root>/wrappers/python (source checkout) or beside
# the example tree (shipped package) -- find it without a hard-coded path
_d = os.path.dirname(os.path.abspath(__file__))
for _ in range(6):
    for _c in (os.path.join(_d, 'wrappers', 'python'), _d):
        if os.path.isfile(os.path.join(_c, 'lumaspdf.py')):
            sys.path.insert(0, _c)
            break
    else:
        _d = os.path.dirname(_d)
        continue
    break
import lumaspdf as L

HERE = os.path.dirname(os.path.abspath(__file__))


def render(template_path, datasets_path, out_pdf_path):
    template = open(template_path, "rb").read()
    datasets = open(datasets_path, "rb").read()
    print("template packet bytes:", len(template))
    print("datasets packet bytes:", len(datasets))

    pdf = L.pdfNewPDF()
    if not pdf:
        print("pdfNewPDF FAILED")
        return -100
    try:
        if L.pdfCreateNewPDFA(pdf, out_pdf_path.encode("latin-1")) == 0:
            print("pdfCreateNewPDFA FAILED")
            return -100

        idx = L.pdfCreateXFAStreamA(pdf, b"template", template, len(template))
        print("pdfCreateXFAStreamA(template) -> index", idx)
        if idx < 0:
            print("pdfCreateXFAStreamA(template) FAILED")
            return -100

        idx = L.pdfCreateXFAStreamA(pdf, b"datasets", datasets, len(datasets))
        print("pdfCreateXFAStreamA(datasets) -> index", idx)
        if idx < 0:
            print("pdfCreateXFAStreamA(datasets) FAILED")
            return -100

        # Phase-6 config gate defaults to enabled (FXFAScriptEnabled := True
        # in Lumas.Pdf.Document.pas's Create), but this driver sets it
        # explicitly anyway so GrandTotalField's <calculate> script is
        # guaranteed to run regardless of that default ever changing -- this
        # example's whole point is the calculate-then-format pipeline.
        prev = L.pdfSetXFAScriptEnabled(pdf, 1)
        print("pdfSetXFAScriptEnabled(1) ->", prev)

        rc = L.pdfRenderXFAForm(pdf)
        print("pdfRenderXFAForm ->", rc)
        if rc < 1:
            print("pdfRenderXFAForm FAILED, code", rc)
            return rc

        if L.pdfCloseFile(pdf) == 0:
            print("pdfCloseFile FAILED")
            return -101

        print("OK: wrote", out_pdf_path)
        return rc
    finally:
        L.pdfDeletePDF(pdf)


def main():
    template_path = os.path.join(HERE, "08_picture_clause_formatting.template.xml")
    datasets_path = os.path.join(HERE, "08_picture_clause_formatting.datasets.xml")
    out_pdf_path = os.path.join(HERE, "08_picture_clause_formatting.pdf")

    print("=== 08_picture_clause_formatting -> 08_picture_clause_formatting.pdf ===")
    rc = render(template_path, datasets_path, out_pdf_path)
    print("RESULT|08_picture_clause_formatting=", rc)

    print()
    print("Expect one page rendered; verify with any text extractor (e.g. pypdf)")
    print("-- the extracted text should read:")
    print("  Purchase Receipt -- Picture-Clause Formatting")
    print("  Customer: Acme Corp")
    print("  Unit Price: 1,875.50")
    print("  Discount: ($125.00)")
    print("  Date: July 24, 2026")
    print("  Phone: 555-123-4567")
    print("  845.25 620.00 410.25")
    print("  Grand Total: 1,875.50")
    print()
    print("  (GrandTotalField has no literal <value> -- its 1,875.50 can only come")
    print("  from the calculate script actually running, THEN the num{zzz,zz9.99}")
    print("  picture formatter being applied to that script's own result.)")


if __name__ == "__main__":
    main()

File: csharp/xfa/08_picture_clause_formatting/08_picture_clause_formatting.cs
// ============================================================================
//  08_picture_clause_formatting -- C# port of examples\delphi\xfa\08_picture_clause_formatting
//  XFA "flavor tour" example 8 of 10: the <format><picture> formatter --
//  real num{}/date{}/text{} picture patterns applied both to plain bound data
//  values and to a value produced by a FormCalc <calculate> script, proving
//  the calculate-then-format pipeline order. A one-page "Purchase Receipt".
//
//  Same pipeline as every example in this tour:
//    pdfNewPDF -> pdfCreateNewPDFA -> pdfCreateXFAStreamA('template',...) ->
//    pdfCreateXFAStreamA('datasets',...) -> pdfRenderXFAForm -> pdfCloseFile
// ============================================================================
using System;
using System.IO;
using System.Runtime.InteropServices;
using LumasPdfSdk;

class PictureClauseFormatting
{
    static int RenderExample(string exeDir, string baseName, string outPdfPath)
    {
        string templatePath = Path.Combine(exeDir, baseName + ".template.xml");
        string datasetsPath = Path.Combine(exeDir, baseName + ".datasets.xml");

        Console.WriteLine("=== " + baseName + " -> " + Path.GetFileName(outPdfPath) + " ===");
        if (!File.Exists(templatePath)) { Console.WriteLine("FILE-NOT-FOUND: " + templatePath); return -100; }

        byte[] templateBuf = File.ReadAllBytes(templatePath);
        byte[] datasetsBuf = File.Exists(datasetsPath) ? File.ReadAllBytes(datasetsPath) : new byte[0];
        Console.WriteLine("template packet bytes: " + templateBuf.Length);
        Console.WriteLine("datasets packet bytes: " + datasetsBuf.Length);

        IntPtr pdf = LumasPdf.pdfNewPDF();
        if (pdf == IntPtr.Zero) { Console.WriteLine("pdfNewPDF FAILED"); return -100; }
        try
        {
            if (!LumasPdf.pdfCreateNewPDFW(pdf, outPdfPath)) { Console.WriteLine("pdfCreateNewPDFA FAILED"); return -100; }

            GCHandle hTpl = GCHandle.Alloc(templateBuf, GCHandleType.Pinned);
            int idx;
            try { idx = LumasPdf.pdfCreateXFAStreamW(pdf, "template", hTpl.AddrOfPinnedObject(), (uint)templateBuf.Length); }
            finally { hTpl.Free(); }
            Console.WriteLine("pdfCreateXFAStreamA(template) -> index " + idx);
            if (idx < 0) { Console.WriteLine("pdfCreateXFAStreamA(template) FAILED"); return -100; }

            if (datasetsBuf.Length > 0)
            {
                GCHandle hDs = GCHandle.Alloc(datasetsBuf, GCHandleType.Pinned);
                try { idx = LumasPdf.pdfCreateXFAStreamW(pdf, "datasets", hDs.AddrOfPinnedObject(), (uint)datasetsBuf.Length); }
                finally { hDs.Free(); }
                Console.WriteLine("pdfCreateXFAStreamA(datasets) -> index " + idx);
                if (idx < 0) { Console.WriteLine("pdfCreateXFAStreamA(datasets) FAILED"); return -100; }
            }
            else
            {
                Console.WriteLine("(no datasets packet found -- template-only render)");
            }

            int rc = LumasPdf.pdfRenderXFAForm(pdf);
            Console.WriteLine("pdfRenderXFAForm -> " + rc + " (expected: page count >= 1)");
            if (rc < 1) { Console.WriteLine("RENDER-FAILED, code " + rc); return rc; }

            if (!LumasPdf.pdfCloseFile(pdf)) { Console.WriteLine("pdfCloseFile FAILED"); return -101; }
            Console.WriteLine("OK: wrote " + outPdfPath + " (" + rc + " page(s))");
            return rc;
        }
        finally
        {
            LumasPdf.pdfDeletePDF(pdf);
        }
    }

    static void Main()
    {
        try
        {
            string exeDir = AppDomain.CurrentDomain.BaseDirectory;
            int rc = RenderExample(exeDir, "08_picture_clause_formatting", Path.Combine(exeDir, "08_picture_clause_formatting.pdf"));
            Console.WriteLine();
            Console.WriteLine("RESULT|08_picture_clause_formatting=" + rc);
            if (rc >= 1)
            {
                Console.WriteLine("OK: rendered " + rc + " page(s). Expect extracted text to read:");
                Console.WriteLine("  Purchase Receipt -- Picture-Clause Formatting");
                Console.WriteLine("  Customer: Acme Corp");
                Console.WriteLine("  Unit Price: 1,875.50            (num{zzz,zz9.99} on \"1875.50\")");
                Console.WriteLine("  Discount: ($125.00)             (num{($zzz,zz9.99)} on \"-125.00\")");
                Console.WriteLine("  Date: July 24, 2026             (date{MMMM DD, YYYY} on \"2026-07-24\")");
                Console.WriteLine("  Phone: 555-123-4567             (text{999-999-9999} on \"5551234567\")");
                Console.WriteLine("  845.25 620.00 410.25            (calc inputs, unformatted)");
                Console.WriteLine("  Grand Total: 1,875.50           (calculate: sum of the three, THEN num{} applied)");
            }
            else
            {
                Console.WriteLine("FAILED, see errors above.");
            }
        }
        catch (Exception e)
        {
            Console.WriteLine("EXCEPTION: " + e.GetType().Name + ": " + e.Message);
        }
    }
}
